presto-spi/src/main/java/com/facebook/presto/spi/block/ColumnarMap.java

```diff
@@ -0,0 +1,160 @@
+            offsets[i] = offsets[i - 1] + (columnarMap.getLength(dictionaryId) * 2);
+        }
+
+        // reindex dictionary
+        int[] dictionaryIds = new int[offsets[dictionaryBlock.getPositionCount()] / 2];
+        int nextDictionaryIndex = 0;
+        for (int i = 1; i < offsets.length; i++) {
+            int dictionaryId = dictionaryBlock.getId(i - 1);
+            int length = columnarMap.getLength(dictionaryId);
+
+            int startOffset = columnarMap.getOffset(dictionaryId);
+            for (int entryIndex = 0; entryIndex < length; entryIndex++) {
+                dictionaryIds[nextDictionaryIndex] = startOffset + entryIndex;
+                nextDictionaryIndex++;
+            }
+        }
+
+        return new ColumnarMap(
+                dictionaryBlock,
+                0,
+                offsets,
+                new DictionaryBlock(dictionaryIds.length, columnarMap.getNullSuppressedKeysBlock(), dictionaryIds),
+                new DictionaryBlock(dictionaryIds.length, columnarMap.getNullSuppressedValuesBlock(), dictionaryIds));
+    }
+
+    private static ColumnarMap toColumnarMap(RunLengthEncodedBlock rleBlock)
+    {
+        ColumnarMap columnarMap = toColumnarMap(rleBlock.getValue());
+
+        // build new offsets block
+        int[] offsets = new int[rleBlock.getPositionCount() + 1];
+        int valueLength = columnarMap.getLength(0);
+        for (int i = 0; i < offsets.length; i++) {
+            offsets[i] = i * valueLength * 2;
+        }
+
+        // create indexes for a dictionary block of the elements
+        int[] dictionaryIds = new int[rleBlock.getPositionCount() * valueLength];
+        int nextDictionaryIndex = 0;
+        for (int i = 1; i < offsets.length; i++) {
+            for (int entryIndex = 0; entryIndex < valueLength; entryIndex++) {
+                dictionaryIds[nextDictionaryIndex] = entryIndex;
+                nextDictionaryIndex++;
+            }
+        }
+
+        return new ColumnarMap(
+                rleBlock,
+                0,
+                offsets,
+                new DictionaryBlock(dictionaryIds.length, columnarMap.getNullSuppressedKeysBlock(), dictionaryIds),
+                new DictionaryBlock(dictionaryIds.length, columnarMap.getNullSuppressedValuesBlock(), dictionaryIds));
+    }
+
+    private ColumnarMap(Block nullCheckBlock, int offsetsOffset, int[] offsets, Block nullSuppressedKeysBlock, Block nullSuppressedValuesBlock)
+    {
+        this.nullCheckBlock = nullCheckBlock;
+        this.offsetsOffset = offsetsOffset;
+        this.offsets = offsets;
+        this.nullSuppressedKeysBlock = nullSuppressedKeysBlock;
+        this.nullSuppressedValuesBlock = nullSuppressedValuesBlock;
+    }
+
+    public int getPositionCount()
+    {
+        return nullCheckBlock.getPositionCount();
+    }
+
+    public boolean isNull(int position)
+    {
+        return nullCheckBlock.isNull(position);
+    }
+
+    public int getLength(int position)
```

Comment:
Maybe call it `entryCount` to avoid confusion about whether the result is 2 or 1 for a single-entry map.